### ISSUE TITLE
Authentication when adding new contacts

### ISSUE THREAD

**Issue by tux3** [qTox/qTox#1063]
Right now, when adding a new contact, one can use a Tox DNS protocols, which are fundamentally insecure due to using DNS, or use raw Tox IDs. Tox IDs are not easy for humans to compare manually, and exchanging those over an insecure channel is really just asking for a MitM. All of this provides plenty of opportunities for an attacker to attempt to impersonate a Tox user.

This is a proposal for providing authentication right after adding someone on Tox. For example, asking a question that only the two peers know the answer to and checking whether their answers match, if they don't, at least one of them might be an impostor.

Doing the authentication with a zero-knowledge proof of equality of the secret answers would give an extremely high probability of authenticity while disallowing any adversary from learning anything about either peer's secrets, and influencing the protocol in any way except making it fail.

**Comment by ArchangeGabriel** [comment]
Related: https://github.com/irungentoo/toxcore/issues/1180, https://github.com/irungentoo/toxcore/issues/1222

**Comment by ProMcTagonist** [comment]
It should be noted that average users are not good at asking questions/providing answers that are accurate down to the character, which zero-knowledge would require.

**Comment by linux-modder** [comment]
@ProMcTagonist , while true....several in the community (myself included)  have asked and volunteered to  setup a 'primer' for  users to ensure this method and  many  other  `best "secure" practices`.... The biggest  end-user hurdle is  education (or often the lack of  such),  A great  way  to get  trust , user-base and  educated safe users would to do something like that, which also would NOT require those  helping to  know how to code......Just a thought.

**Comment by ProMcTagonist** [comment]
@linux-modder What sort of "primer" do you mean? An introductory guide? I would love to help.

Sorry for off-topic.

**Comment by linux-modder** [comment]
@ProMcTagonist  Precisely....  I have a  nice  project  planning to make a  secure  laptop that is a true all in one and has 2  501(c)3 's  planned one is  purely  for  educating users  so they don't  just  disable  everything then  bitch they got had...(...../ameridea/SkunkWorks if you're interested)

**Comment by SkyzohKey** [comment]
A nice idea would be to allow users to setup their PGP public key to their profile. If another client request it (via custom packets) qTox would send the PGP pubkey to the other client. Then the other client could verify the user easily.
